testing/x/json_match.go

```diff
@@ -77,6 +77,14 @@ func AssertJSONMatchesPattern(
 				t.Errorf("[%s]: expected an array of length %d, but got %s",
 					path, len(pp), prettyJSON(t, a))
 			}
+
+			sort.SliceStable(aa, func(i, j int) bool {
+				return strings.Compare(
+					fmt.Sprintf("%v", aa[i]),
+					fmt.Sprintf("%v", aa[j]),
+				) < 0
+			})
```

Comment:
I don't think we want to take this change. The arrays `[1, 2]` and `[2, 1]` are not semantically equivalent in most cases. This feels like it is covering for a missing sort somewhere else.

CC @t0yv0 


FYI:
See https://go.dev/src/strings/compare.go
```suggestion
			sort.SliceStable(aa, func(i, j int) bool {
				return fmt.Sprintf("%v", aa[i]) < fmt.Sprintf("%v", aa[j]),
			})
```

Reply:
Yeah this would throw me off as highly unintuitive. 

```
AssertJSONMatchesPattern(t, []byte(`[1, 2, 3]`), []byte(`[1, 3, 2]`))
```

If you go this way you may be interested in growing the pattern language like having some way to express that ```{"containsExactly": [1,2,3]}``` pattern that would  match `[1,2,3]`, `[1,3,2]` and so on.

Building this out though we're straying from core purpose so perhaps instead adopt something like [Gomega](https://onsi.github.io/gomega/) that was recently demonstrated? It might have existing rich vocabulary for this sort of matching.

Reply:
The failures returned in the GRPC responses are not deterministically ordered. I added this change in order to be able to assert on that.

I agree that this doesn't feel like the right thing to build and we should probably adopt some library for this. I'll give Gomega a go

Reply:
The source of the nondeterminism is https://github.com/pulumi/terraform-plugin-sdk/blob/e6d96b3b8f7e3064e39d98976bf6dac158bf8d89/helper/schema/schema.go#L2182

During the validation we iterate over a map and that is nondeterministic in go.

I added a variant of replay which accepts options and the only current option is a map of paths of arrays which are unordered. LMK if that makes sense. I think the current interface of `Replay` is quite nice so I think it's worth keeping. As for Gomega - we could opt to instead implement the matching that way but I think this is out of scope of this PR. Happy to work on that in a follow-up if we have a further use cases.